binary_search_tree/tree.py

```diff
@@ -14,35 +14,128 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+    
+    def add_helper(self, current, key, value):
+        if current == None:
+            return TreeNode(key, value)
+        if current.key > key:
+            current.left = self.add_helper(current.left, key, value)
+        else:
+            current.right = self.add_helper(current.right, key, value)
+        return current  
+        
 
-    # Time Complexity: 
-    # Space Complexity: 
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
```

Comment:
👍  But the space complexity is O(log n) for a balanced tree because you're using recursion.